src/linear_algebra/charpoly.lean

```diff
@@ -0,0 +1,100 @@
+/-
+Copyright (c) 2021 Riccardo Brasca. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Riccardo Brasca
+-/
+
+import linear_algebra.free_module
+import linear_algebra.charpoly.coeff
+import linear_algebra.matrix.basis
+import linear_algebra.dimension
+
+/-!
+
+# Characteristic polynomial
+
+We define the characteristic polynomial of `f : M →ₗ[R] M`, where `M` is a free `R`-module.
+
+## Main definition
+
+* `linear_map.charpoly f` : the characteristic polynomial of `f : M →ₗ[R] M`.
+
+-/
+
+universes u v w
+
+variables {R : Type u} {M : Type v} [comm_ring R] [nontrivial R]
+variables [add_comm_group M] [module R M] [module.free R M] [module.finite R M] (f : M →ₗ[R] M)
+
+open_locale classical
+
+noncomputable theory
+
+open module.free polynomial matrix
+
+namespace linear_map
+
+section basic
+
+/-- The characteristic polynomial of `f : M →ₗ[R] M`. -/
+def charpoly := (linear_map.to_matrix (choose_basis R M) (choose_basis R M) f).charpoly
+
+lemma charpoly_def :
+  f.charpoly = (linear_map.to_matrix (choose_basis R M) (choose_basis R M) f).charpoly := rfl
+
+/-- `charpoly f` is the characteristic polynomial of the matrix of `f` in any basis. -/
+lemma charpoly_eq_matrix.charpoly {ι : Type w} [fintype ι] (b : basis ι R M) :
+  f.charpoly = (linear_map.to_matrix b b f).charpoly :=
+begin
+  set b' := (choose_basis R M),
+  set P := reindex_linear_equiv R R (equiv_of_basis b' b) (equiv.refl ι) (b'.to_matrix b),
+  set Q := reindex_linear_equiv R R (equiv.refl ι) (equiv_of_basis b' b) (b.to_matrix b'),
+
+  have hPQ : (P.map C).mul (Q.map C) = 1,
+  { rw [← matrix.map_mul, ← reindex_linear_equiv_mul R R, basis.to_matrix_mul_to_matrix_flip,
+      reindex_linear_equiv_one, ← ring_hom.map_matrix_apply, ring_hom.map_one],
+    apply_instance },
```

Comment:
I don't know why Lean doesn't find the instance by itself.

Reply:
Basically because `rw` has to elaborate `reindex_linear_equiv_mul R R` before it can search for instances of `reindex_linear_equiv _ _ _ _ (mul _ _)`, and in this elaboration process it can't yet figure out that the second parameter of `reindex_linear_equiv_mul` has to be `ι`, therefore it can't find the instance. You could rewrite using `@reindex_linear_equiv_mul _ ι _ _ _ _ R R` instead, but the `apply_instance` is fine for me too.